## WF2 Module 2

We want to find molecular phenotypes that are integral to a disease, and then chemicals that activate those phenotypes, and then genes that clean up those bad chemicals.   The tricky part with current data is that it's difficult to make the jump to chemicals.

For instance, if we say that molecular phenotype is a GO process, we might find that e.g. various DNA repair processes are involved with Fanconi Anemia.  But the only chemicals that are linked to these processes are participants in the process, like DNA.

From the chemical side, there are properties on chemicals in CHEBI, such as being genotoxic.  However, there's not any semantic value to these properties because they are just strings with no relation to e.g. GO terms.  This gap is being filled by CHIRO, but that is still ongoing.

To leapfrog this, we'll try (1) grabbing chemicals that look bad for the disease, (2) try to do some enrichments on their properties to get other similar chemicals, and (3) look for genes that degrade those things.

In [1]:
#To make nicer looking outputs
from IPython.core.display import display, HTML
import requests
import pandas as pd
import os
import sys

#Load some functions for parsing quick output
module_path = os.path.abspath(os.path.join('../..'))
if module_path not in sys.path:
        sys.path.append(module_path)
from gg_functions import parse_answer, get_view_url, expand, quick

robokop='robokop.renci.org' 

In [2]:
disease = "MONDO:0019391"  #Fanconi Anemia

In [3]:
bad_chemicals = expand('disease',disease,'chemical_substance',predicate='contributes_to',direction='in')

In [4]:
df = parse_answer(bad_chemicals,node_list=['n1'])

In [5]:
df

n1 - id      n1 - name     score
0  CHEBI:23704  diepoxybutane  0.513305
1  CHEBI:16856    glutathione  0.508893
2  CHEBI:28901       busulfan  0.360132

What properties do these things have in common?

In [6]:
chems = [x for x in bad_chemicals['knowledge_graph']['nodes'] if 'chemical_substance' in x['type']]
from collections import defaultdict
prop_counts = defaultdict(int)
for n in chems:
    for p in n:
        prop_counts[p] += 1
items = list(prop_counts.items())
count_frame = pd.DataFrame.from_records(items,columns=['property','count_in_subset'])
count_frame.head()

property  count_in_subset
0  eukaryotic_metabolite                1
1       therapeutic_flag                3
2                topical                3
3            application                2
4                   mass                3

Many of these propertries are assigned to many chemicals (like SMILES or name).  We want to find the ones that are overrepresented.  To do that we need to know the details of how often each property is used in the database.  At the moment, ROBOKOP does not expose functions to allow this, but we have pre-calculated such a list:

In [7]:
property_counts = pd.read_csv('../../examples/chemprops.txt','\t')
property_counts.head()

property  count
0                                  appetite_enhancer      2
1         Wiskott_Aldrich_syndrome_protein_inhibitor      1
2                           peptide_coupling_reagent      2
3                 EC_3.4.22.52_(calpain_1)_inhibitor      3
4  EC_2.3.1.5_(arylamine_N_acetyltransferase)_inh...      1

In [8]:
df = pd.merge(count_frame, property_counts,on='property',how='inner')

df.head()

property  count_in_subset   count
0  eukaryotic_metabolite                1   11711
1       therapeutic_flag                3  244803
2                topical                3  244803
3            application                2    8162
4                   mass                3   97320

In [9]:
from scipy.stats import hypergeom
#in scipy:
#The hypergeometric distribution models drawing objects from a bin. 
#M is the total number of objects, 
#n is total number of Type I objects. 
#The random variate represents the number of Type I objects in N drawn without replacement from the total population.

total_chemical_count=355413 # M above
subset_count = len(chems) # N above

def calc_enrich_p(x,n):
    return hypergeom.sf(x-1, total_chemical_count, n, subset_count)

In [10]:
df['enrichment_p'] = df.apply(lambda x: calc_enrich_p(x['count_in_subset'], x['count']), axis=1)
df.sort_values(by='enrichment_p')

property  count_in_subset   count  enrichment_p
44                      mutagen                2     195  8.981212e-07
42                    genotoxin                2     237  1.327776e-06
45       aetiopathogenetic_role                2     835  1.651312e-05
47             insect_sterilant                1       7  5.908519e-05
51               chemosterilant                1       7  5.908519e-05
13        skin_lightening_agent                1       7  5.908519e-05
18                     cosmetic                1      12  1.012875e-04
56            teratogenic_agent                1      47  3.966702e-04
39              biological_role                3   27273  4.518088e-04
52             alkylating_agent                1      78  6.582463e-04
3                   application                2    8162  1.557747e-03
53           carcinogenic_agent                1     256  2.159316e-03
49                  insecticide                1     387  3.263076e-03
24                     cofactor                1     468  3.945145e-03
15                  antioxidant                1     498  4.197685e-03
43                       entity                2   13726  4.358989e-03
41                         role                2   13726  4.358989e-03
46                   continuant                2   13726  4.358989e-03
48                    pesticide                1    1105  9.298235e-03
10  Escherichia_coli_metabolite                1    1299  1.092471e-02
55         antineoplastic_agent                1    1546  1.299296e-02
4                          mass                3   97320  2.053036e-02
8             monoisotopic_mass                3   98074  2.101126e-02
9          bacterial_metabolite                1    2513  2.106237e-02
12       prokaryotic_metabolite                1    2528  2.118720e-02
34                       charge                3   98675  2.139991e-02
21             mouse_metabolite                1    3083  2.579823e-02
19             human_metabolite                1    4637  3.863205e-02
54                         drug                1    5527  4.593116e-02
50               pharmaceutical                1    5605  4.656909e-02
5          mammalian_metabolite                1    6871  5.688347e-02
17            animal_metabolite                1    7438  6.147872e-02
32                Bronsted_base                1    7524  6.217440e-02
20                         base                1    7531  6.223101e-02
38                     acceptor                1    7542  6.231996e-02
0         eukaryotic_metabolite                1   11711  9.563004e-02
40                chemical_role                1   16524  1.330934e-01
25                   metabolite                1   20776  1.653167e-01
7              biochemical_role                1   22692  1.795720e-01
28          molecule_properties                3  244803  3.267749e-01
1              therapeutic_flag                3  244803  3.267749e-01
33                         oral                3  244803  3.267749e-01
31                   parenteral                3  244803  3.267749e-01
2                       topical                3  244803  3.267749e-01
35               withdrawn_flag                3  244803  3.267749e-01
22              natural_product                3  244803  3.267749e-01
14                molecule_type                3  244803  3.267749e-01
16                      prodrug                3  244820  3.268429e-01
11                   iupac_name                3  311601  6.739010e-01
6                          name                3  311887  6.757583e-01
27             molecular_weight                3  312022  6.766362e-01
36                     inchikey                3  315397  6.988312e-01
26                        inchi                3  315397  6.988312e-01
30                       smiles                3  320976  7.365757e-01
37            molecular_formula                3  324167  7.587630e-01
29       equivalent_identifiers                

Now we can see that the most enriched property is "mutagen", which occurs in 2 of the 3 chemicals that are contributors to FA, but in only 195 out of the 355k chemicals.  Let's look for genes that degrade mutagens.  We're going to lose the connection to FA, so that won't be included in our score, which is unfortunate.

In [11]:
def create_question(disease_id,prop):
    return {
    "machine_question": {
        "nodes": [
            {
                "id": "n0",
                "type": "chemical_substance",
                prop: True
            },
            {
                "id": "n1",
                "type": "gene"
            },
            {
                "id": "n2",
                "type": "disease",
                "curie": disease_id
            }
        ],
        "edges": [
            {
                "id": "e0",
                "source_id": "n1",
                "target_id": "n0",
                "type": "increases_degradation_of"
            }
        ]
    }
}

In [12]:
q = create_question('MONDO:0019391','mutagen')
a = quick(q)

Return Status: 200


In [13]:
aframe = parse_answer(a,node_list=['n0','n1'])

In [14]:
aframe

n0 - id               n0 - name     n1 - id n1 - name     score
0   CHEBI:15343            acetaldehyde    HGNC:404     ALDH2  0.263911
1   CHEBI:15343            acetaldehyde    HGNC:405   ALDH3A1  0.257250
2   CHEBI:27732                caffeine   HGNC:2596    CYP1A2  0.190395
3    CHEBI:4027        cyclophosphamide   HGNC:2637    CYP3A4  0.183836
4   CHEBI:17754                glycerol   HGNC:4289        GK  0.181862
5   CHEBI:15343            acetaldehyde    HGNC:407   ALDH1B1  0.181362
6   CHEBI:15343            acetaldehyde    HGNC:877   ALDH7A1  0.181250
7   CHEBI:15343            acetaldehyde    HGNC:409   ALDH1A3  0.181236
8   CHEBI:15343            acetaldehyde    HGNC:412   ALDH9A1  0.181208
9   CHEBI:15343            acetaldehyde    HGNC:410   ALDH3B1  0.181194
10   CHEBI:2560         Aldophosphamide    HGNC:405   ALDH3A1  0.180767
11  CHEBI:28534       1,2-dibromoethane   HGNC:4641     GSTT1  0.180766
12  CHEBI:17754                glycerol  HGNC:28635       GK5  0.180746
13  CHEBI:28262      dimethyl sulfoxide   HGNC:4289        GK  0.180741
14  CHEBI:17754                glycerol   HGNC:4291       GK2  0.180731
15  CHEBI:28262      dimethyl sulfoxide    HGNC:380    AKR1A1  0.180731
16  CHEBI:17754                glycerol    HGNC:381    AKR1B1  0.180727
17  CHEBI:17754                glycerol    HGNC:382   AKR1B10  0.180727
18  CHEBI:15343            acetaldehyde    HGNC:403   ALDH3A2  0.145771
19  CHEBI:15343            acetaldehyde    HGNC:411   ALDH3B2  0.145771
20  CHEBI:76565  2-ammonioprop-2-enoate  HGNC:10691       SDS  0.145505
21  CHEBI:17123     2-aminoacrylic acid  HGNC:10691       SDS  0.145505
22   CHEBI:2560         Aldophosphamide   HGNC:4638     GSTP1  0.145426
23  CHEBI:28534       1,2-dibromoethane   HGNC:4638     GSTP1  0.145426
24   CHEBI:2560         Aldophosphamide   HGNC:4641     GSTT1  0.145367
25  CHEBI:28534       1,2-dibromoethane   HGNC:4632     GSTM1  0.145365
26   CHEBI:2560         Aldophosphamide   HGNC:4632     GSTM1  0.145365
27  CHEBI:80559         Alcophosphamide    HGNC:253      ADH5  0.145354
28  CHEBI:28534       1,2-dibromoethane   HGNC:4627     GSTA2  0.145354
29   CHEBI:2560         Aldophosphamide   HGNC:4627     GSTA2  0.145354
..          ...                     ...         ...       ...       ...
64  CHEBI:28534       1,2-dibromoethane   HGNC:4636     GSTM4  0.087940
65   CHEBI:2560         Aldophosphamide   HGNC:7064     MGST3  0.087940
66  CHEBI:84254   N,N-dimethylacetamide    HGNC:380    AKR1A1  0.087940
67   CHEBI:2560         Aldophosphamide  HGNC:19662     GSTA5  0.087940
68  CHEBI:28534       1,2-dibromoethane   HGNC:4629     GSTA4  0.087940
69  CHEBI:80559         Alcophosphamide    HGNC:251     ADH1C  0.087940
70   CHEBI:2560         Aldophosphamide  HGNC:33437    GSTT2B  0.087940
71  CHEBI:28534       1,2-dibromoethane  HGNC:13312     GSTO1  0.087940
72   CHEBI:2560         Aldophosphamide   HGNC:4628     GSTA3  0.087940
73  CHEBI:84254   N,N-dimethylacetamide    HGNC:381    AKR1B1  0.087940
74   CHEBI:2560         Aldophosphamide   HGNC:4635     GSTM3  0.087940
75  CHEBI:80559         Alcophosphamide    HGNC:256      ADH7  0.087940
76  CHEBI:80559         Alcophosphamide    HGNC:255      ADH6  0.087940
77  CHEBI:80559         Alcophosphamide    HGNC:250     ADH1B  0.087940
78   CHEBI:2560         Aldophosphamide   HGNC:4637     GSTM5  0.087940
79  CHEBI:84254   N,N-dimethylacetamide   HGNC:4289        GK  0.087940
80  CHEBI:28534       1,2-dibromoethane   HGNC:4635     GSTM3  0.087940
81  CHEBI:28534       1,2-dibromoethane  HGNC:19662     GSTA5  0.087940
82  CHEBI:80559         Alcophosphamide    HGNC:249     ADH1A  0.087940
83  CHEBI:84254   N,N-dimethylacetamide   HGNC:4291       GK2  0.087940
84  CHEBI:76565  2-ammonioprop-2-enoate  HGNC:30404      SDSL  0.087940
85  CHEBI:80559         Alcophosphamide    HGNC:252      ADH4  0.087940
86   CHEBI:2560         Aldophosphamide    HGNC:411   ALDH3B2  0.087940
87  CHEBI:17123     2-aminoacrylic acid  HGNC:30

This finds the expected result that ALDH2 is a potential modifier for FA.   Potentially, but looking for things that degrade into acetaldehyde or other chemicals in n0 we could find a list of environemntal exposures to avoid.  Note that, as ever, we are at the mercy of the data - caffeine is mutagenic at high dose but mostly in non-mammalian cells.